source/common/network/connection_impl.cc

```diff
@@ -135,10 +135,21 @@ void ConnectionImpl::close(ConnectionCloseType type) {
     return;
   }
 
+  if (type == ConnectionCloseType::HalfClose && remote_half_closed_) {
+    // If both sides have half-closed, proceed to fully close the
+    // connection after flushing.
+    type = ConnectionCloseType::FlushWrite;
+  }
+
   uint64_t data_to_write = write_buffer_->length();
   ENVOY_CONN_LOG(debug, "closing data_to_write={} type={}", *this, data_to_write, enumToInt(type));
-  if (data_to_write == 0 || type == ConnectionCloseType::NoFlush ||
-      !transport_socket_->canFlushClose()) {
+  if (type == ConnectionCloseType::HalfClose) {
+    local_half_closed_ = true;
```

Comment:
Can you verify a corner case for me?   Let's assume we get a giant read with a FIN.  So we set local_half_closed_ true but the write results in backup.     When does the fin actually get sent?  

I'd think you need something akin to close_with_flush_ where we'd mark should_send_fin_ here and then local_half_closed_ or sent_fin_ once we've flushed the write.

Reply:
Good catch.